sharded_queue/router/tube.lua

```diff
@@ -0,0 +1,415 @@
+    end
+
+    return task
+end
+
+function take(self, timeout, options)
+    options = table.deepcopy(options or {})
+    options.tube_name = self.tube_name
+
+    options.extra = {
+        log_request = utils.normalize.log_request(options.log_request) or self.log_request,
+    }
+
+    local remote_call_timeout = time.MIN_NET_BOX_CALL_TIMEOUT
+    if timeout ~= nil and timeout > time.MIN_NET_BOX_CALL_TIMEOUT then
+        remote_call_timeout = timeout
+    end
+
+    local take_timeout = time.nano(timeout) or time.TIMEOUT_INFINITY
+
+    local frequency = 1000
+    local wait_part = 0.01 -- maximum waiting time in second
+    local wait_max = utils.normalize.wait_max(options.wait_max)
+        or self.wait_max or time.MAX_TIMEOUT
+
+    local wait_factor = self.wait_factor
+
+    local calc_part = time.sec(take_timeout / frequency)
+
+    if calc_part < wait_part then
+        wait_part = tonumber(calc_part)
+    end
+
+    if options.extra.log_request then
+        log.info(("Router[%d] take: start attempts"):format(fiber.self():id()))
+    end
+
+    while take_timeout ~= 0 do
+        local begin = time.cur()
+
+        local shards, err = vshard.router.routeall()
+        if err ~= nil then
+            error(err)
+        end
+
+        local replicasets = {}
+        for _, replicaset in pairs(shards) do
+            table.insert(replicasets, replicaset)
+        end
+        utils.array_shuffle(replicasets)
+
+        local task = take_task(replicasets,
+            options, take_timeout, remote_call_timeout)
+
+        if task ~= nil then
+            if options.extra.log_request then
+                log_task('take', task)
+            end
+            return task
+        end
+
+        if take_timeout < time.nano(wait_part) then
+            if options.extra.log_request then
+                log.info(("Router[%d] take: next attemt will be after timeout")
+                    :format(fiber.self():id()))
+            end
+            return nil
+        end
+
+        fiber.sleep(wait_part)
+
+        wait_part = wait_part * wait_factor
+        if wait_part > wait_max then
+            wait_part = wait_max
+        end
+
+        local duration = time.cur() - begin
+
+        take_timeout = take_timeout > duration and take_timeout - duration or 0
+    end
+
+    if options.extra.log_request then
+        log.info(("Router[%d] take: timeout"):format(fiber.self():id()))
+    end
+end
+
+function delete(self, task_id, options)
+    local bucket_count = vshard.router.bucket_count()
+    local bucket_id, _ = utils.unpack_task_id(task_id, bucket_count)
+
+    options = table.deepcopy(options or {})
+    options.tube_name = self.tube_name
+    options.task_id = task_id
+
+    options.extra = {
+        log_request = utils.normalize.log_request(options.log_request) or self.log_request,
+    }
+
+    local task, err = vshard.router.call(bucket_id, 'write', 'tube_delete', {
+        options
+    })
+    -- re-raise storage errors
```

Comment:
```suggestion
    -- Re-raise storage errors.
``` 
And so on. There is a bunch of copies of this comment. Up to you.

Reply:
Thank you, updated.